File: voan/plan.py
"""Plan-then-execute — the architectural tier that beats in-domain hijacks.

The judge asks "is this action consistent with the goal?" — which it cannot answer
for an in-domain action (a send_money that pays a bill vs. one that pays an
attacker look identical without provenance). Plan-then-execute sidesteps that: the
agent commits the list of actions it INTENDS to take *before* it reads any
untrusted data, and Voan then allows ONLY those actions. An instruction injected
into a tool result cannot add a new action, or change the target of a planned one,
because the plan was fixed before the poison was ever seen. (Pattern from "Design
Patterns for Securing LLM Agents against Prompt Injections", Beurer-Kellner 2025.)

A Plan is a multiset of PlanSteps consumed as the agent runs. A step is
(tool, pins): `pins` are argument values that MUST match (pin the recipient of a
payment and an injected recipient-swap is rejected); an empty `pins` matches any
args for that tool. First unused matching step wins and is consumed, so N planned
calls authorize exactly N executions — an injected extra call finds no step and is
blocked.
"""


class PlanStep:
    __slots__ = ("tool", "pins", "used")

    def __init__(self, tool, pins=None):
        self.tool = tool
        self.pins = pins or {}
        self.used = False

    def matches(self, action):
        if self.used or self.tool != action.tool:
            return False
        args = action.args if isinstance(action.args, dict) else {}
        return all(_eq(args.get(k), v) for k, v in self.pins.items())


class Plan:
    """An ordered list of intended steps, consumed one per matching action."""

    def __init__(self, steps):
        self.steps = [_step(s) for s in steps]

    def allows(self, action):
        """True (and consume the step) if `action` was planned; else False."""
        for s in self.steps:
            if s.matches(action):
                s.used = True
                return True
        return False

    def reset(self):
        for s in self.steps:
            s.used = False
        return self


def _step(s):
    if isinstance(s, PlanStep):
        return s
    if isinstance(s, str):
        return PlanStep(s)
    if isinstance(s, dict):
        # {"tool": "send_money", "pins": {"recipient": "GB.."}}  or
        # {"tool": "send_money", "recipient": "GB.."}  (bare pins)
        tool = s.get("tool")
        pins = s.get("pins")
        if pins is None:
            pins = {k: v for k, v in s.items() if k != "tool"}
        return PlanStep(tool, pins)
    if isinstance(s, (list, tuple)):        # (tool, pins)
        return PlanStep(s[0], s[1] if len(s) > 1 else None)
    raise TypeError(f"cannot read plan step: {s!r}")


def _eq(a, b):
    return str(a).strip().lower() == str(b).strip().lower()

File: sdk-js/src/plan.ts
// Plan-then-execute — the architectural tier that beats in-domain hijacks.
// Mirror of voan/plan.py. The agent commits the actions it INTENDS to take before
// it reads untrusted data; Voan then allows ONLY those (each consumed once, args
// optionally pinned), so an injection can't add a new action or swap a planned
// one's recipient. Deterministic — no LLM. (Design Patterns for Securing LLM
// Agents against Prompt Injections, Beurer-Kellner 2025.)
import type { Action } from "./schema.ts";

export interface PlanStepInit {
  tool: string;
  pins?: Record<string, unknown>;
  [arg: string]: unknown; // bare pins: { tool: "send_money", recipient: "GB.." }
}

export type PlanStepSpec = string | PlanStepInit | [string, Record<string, unknown>?];

export class PlanStep {
  tool: string;
  pins: Record<string, unknown>;
  used = false;

  constructor(tool: string, pins?: Record<string, unknown>) {
    this.tool = tool;
    this.pins = pins ?? {};
  }

  matches(action: Action): boolean {
    if (this.used || this.tool !== action.tool) return false;
    const args = (action.args && typeof action.args === "object") ? action.args : {};
    return Object.entries(this.pins).every(([k, v]) => eq((args as Record<string, unknown>)[k], v));
  }
}

export class Plan {
  steps: PlanStep[];
  constructor(steps: PlanStepSpec[]) {
    this.steps = steps.map(toStep);
  }

  /** True (and consume the step) if `action` was planned; else false. */
  allows(action: Action): boolean {
    for (const s of this.steps) {
      if (s.matches(action)) { s.used = true; return true; }
    }
    return false;
  }

  reset(): this {
    for (const s of this.steps) s.used = false;
    return this;
  }
}

function toStep(s: PlanStepSpec): PlanStep {
  if (typeof s === "string") return new PlanStep(s);
  if (Array.isArray(s)) return new PlanStep(s[0], s[1]);
  const { tool, pins, ...rest } = s;
  return new PlanStep(tool, pins ?? (Object.keys(rest).length ? rest : undefined));
}

function eq(a: unknown, b: unknown): boolean {
  return String(a).trim().toLowerCase() === String(b).trim().toLowerCase();
}
